docs(splunk service): added highlight article for Splunk HEC improvements (#10473)
vectordotdev/vector: website/content/en/highlights/2021-12-15-splunk-hec-improvements.md

---
date: "2021-12-15"
title: "Splunk HEC Improvements"
description: "Improved compatibility with Splunk HTTP Event Collector"
authors: ["barieom"]
pr_numbers: [10261, 10135]
release: "0.19.0"
hide_on_release_notes: false
badges:
  type: announcement
---

We're excited to share that we've made a couple improvements to Vector's support

> for Splunk HEC; Vector now supports Splunk HEC indexer acknowledgements and

Reviewer: I might add some links to the Splunk HEC docs for people who are familiar with Splunk, but not the HEC format.

token passthrough routing.

> With the new support for Splunk HEC indexer acknowledgement, Vector now provides a

Reviewer 2: ```suggestion
## Indexer Acknowledgements

With the new support for Splunk HEC indexer acknowledgement, Vector now provides a
```

better end-to-end acknowledgement to ensure that no data is lost. As you may be
aware, Splunk HEC does not guarantee that a data is successfully written and saved

> if it receives a 200. Vector will now check [Splunk HEC indexer acknowledgements][Splunk indexer]

Reviewer: ```suggestion
if it responds to the incoming request successfully. Vector will now check [Splunk HEC indexer acknowledgements][Splunk indexer]
```

protocol to verify that data has been successfully delivered. To learn more about

> how it works, check out the short explanation in the Vector [sinks reference page][indexer how it works].

Reviewer: ```suggestion
how it works, check out the short explanation in the Vector [`splunk_hec` sink reference page][indexer how it works].
```

> Vector now also supports new token passthrough routing. You can now use these tokens,

Reviewer 2: ```suggestion
## Passthrough Token Routing

Vector now also supports new token passthrough routing. You can now use these tokens,
```

which can be found in `splunk_hec` source requests, downstream `splunk_hec` sink
requests. With the ability to record Splunk token that the events came in with, you

> can now use that token in the `splunk_hec` sink to forward logs. With this update,

Reviewer 3: This seems slightly repetitive with the previous sentence. Perhaps combine them into something like: 
> "When `store_hec_token` is enabled in a `splunk_hec` source, tokens included in requests to the source will be stored and used by downstream `splunk_hec` sinks. Any passed through token takes precedence over the `default_token` configuration set in the sink." 

Side note: the token passthrough works for both `splunk_hec` metrics and logs sinks so "to foward logs" may be a bit confusing.

> you can now partition event batches in the `splunk_hec` sinks by token.

Reviewer 3: I think you can leave this partitioning bit out as it's more of a minor implementation detail.

---
date: "2021-12-15"
title: "Splunk HEC Improvements"
description: "Improved compatibility with Splunk HTTP Event Collector"
authors: ["barieom"]
pr_numbers: [10261, 10135]
release: "0.19.0"
hide_on_release_notes: false
badges:
  type: enhancement
---

We're excited to share that we've made a few improvements to Vector's support
for [Splunk HEC][Splunk HEC]; Vector now supports Splunk HEC indexer
acknowledgements and channel token passthrough routing.

## Indexer Acknowledgements

With the new support for Splunk HEC indexer acknowledgement, Vector is now able
to provide higher guarantees that no data is lost when using the `splunk_hec`
source and sink.

As you may be aware, Splunk HEC does not guarantee that data is successfully

> written when responds to the incoming request successfully, but rather specifies
> that the acknowledgement status of sent events is provided via a [separate
> endpoint][indexer].

Reviewer 3: ```suggestion
written when it responds to an incoming request successfully. To confirm that data has been persisted, the [indexer acknowledgements][indexer] feature must be enabled. The status of sent events can then be queried via a separate endpoint.
```

### `splunk_hec` source acknowledgements

Previously, the `splunk_hec`  source did not support [indexer
acknowledgements][indexer] and so would provide weaker delivery guarantees and
would not work with Splunk senders that required them.

> Now, you can configure the `splunk_hec` source to responding to indexer
> acknowledgement requests from clients by configuring:

Reviewer 3: ```suggestion
Now, you can configure the `splunk_hec` source to use the indexer acknowledgements protocol by configuring:
```

```toml
acknowledgements = true
```

> In addition to this configuring the source to expose `/services/collector/ack`
> for querying for ack status by clients, this also wires into Vector's

Reviewer 3: ```suggestion
When enabled, responses to incoming requests will include an ID that can be used to query for acknowledgement status at the newly exposed `/services/collector/ack` endpoint ([learn more here][indexer how it works]). The acknowledgement status is wired into Vector's
```

forthcoming end-to-end acknowledgement feature. This feature will require that
events be sent by sinks or persisted into disk buffers before sources will
acknowledge them.

### `splunk_hec` sink acknowledgements

Previously, the `splunk_hec` sink simply treated successful HEC requests as the
events being acknowledged by the Spunk receiver and so Vector would drop them
from any buffers. Now, it is possible to configure the sink to wait until the

> Splunk reciever acknowledges the events via the [index

Reviewer 3: ```suggestion
Splunk receiver acknowledges the events via the [index
```

acknowledgements][indexer] part of the HEC protocol.

This has defaulted to on to provide higher guarantees, but can be disabled to
restore the previous behavior by configuring:

```toml
acknowledgements.indexer_acknowledgements_enabled = false
```

## Passthrough Token Routing

Vector now also supports Splunk HEC token passthrough routing. When
`store_hec_token` is enabled in a `splunk_hec` source, tokens included in
requests to the source will be stored and used by downstream `splunk_hec` sinks.
Any passed through token takes precedence over the `default_token` configuration

> set in the sink. With this update, you can now partition event batches in the
> `splunk_hec` sinks by token.

Reviewer 3: ```suggestion
set in the sink.
```
